pfcp/adapter/adapter.go

```diff
@@ -149,128 +169,180 @@ func HandlePfcpAssociationSetupResponse(msg *pfcpUdp.Message) {
-		if eventData, ok := msg.EventData.(pfcpUdp.PfcpEventData); !ok {
+		if eventData, ok := msg.EventData.(udp.PfcpEventData); !ok {
 			logger.PfcpLog.Warnf("PFCP Session Establish Response found invalid event data, response discarded")
 			return
 		} else {
 			SEID = eventData.LSEID
 		}
 	}
-	smContext := smf_context.GetSMContextBySEID(SEID)
+	smContext := context.GetSMContextBySEID(SEID)
 	logger.PfcpLog.Infof("In HandlePfcpSessionEstablishmentResponse SEID %v", SEID)
 	logger.PfcpLog.Infof("In HandlePfcpSessionEstablishmentResponse smContext %+v", smContext)
 
 	// Get NodeId from Seq:NodeId Map
-	seq := msg.PfcpMessage.Header.SequenceNumber
+	seq := rsp.Sequence()
 	nodeID := FetchPfcpTxn(seq)
 
 	if rsp.UPFSEID != nil {
 		NodeIDtoIP := nodeID.ResolveNodeIdToIp().String()
 		pfcpSessionCtx := smContext.PFCPContext[NodeIDtoIP]
-		pfcpSessionCtx.RemoteSEID = rsp.UPFSEID.Seid
-		smContext.SubPfcpLog.Infof("in HandlePfcpSessionEstablishmentResponse rsp.UPFSEID.Seid [%v] ", rsp.UPFSEID.Seid)
+		rspUPFseid, err := rsp.UPFSEID.FSEID()
+		if err != nil {
+			logger.PfcpLog.Errorf("pfcp session establishment response UPFSEID error: %v", err)
+			return
+		}
+		pfcpSessionCtx.RemoteSEID = rspUPFseid.SEID
+		smContext.SubPfcpLog.Infof("in HandlePfcpSessionEstablishmentResponse rsp.UPFSEID.Seid [%v] ", rspUPFseid.SEID)
 	}
 
 	// Get N3 interface UPF
 	ANUPF := smContext.Tunnel.DataPathPool.GetDefaultPath().FirstDPNode
 
 	if ANUPF.UPF.NodeID.ResolveNodeIdToIp().Equal(nodeID.ResolveNodeIdToIp()) {
+		if rsp.Cause == nil {
+			logger.PfcpLog.Errorln("pfcp session establishment response has no cause")
+			return
+		}
+		causeValue, err := rsp.Cause.Cause()
+		if err != nil {
+			logger.PfcpLog.Errorf("pfcp session establishment response cause error: %v", err)
+			return
+		}
 		// UPF Accept
-		if rsp.Cause.CauseValue == pfcpType.CauseRequestAccepted {
-			smContext.SBIPFCPCommunicationChan <- smf_context.SessionEstablishSuccess
+		if causeValue == ie.CauseRequestAccepted {
+			smContext.SBIPFCPCommunicationChan <- context.SessionEstablishSuccess
 			smContext.SubPfcpLog.Infof("PFCP Session Establishment accepted")
 		} else {
-			smContext.SBIPFCPCommunicationChan <- smf_context.SessionEstablishFailed
-			smContext.SubPfcpLog.Errorf("PFCP Session Establishment rejected with cause [%v]", rsp.Cause.CauseValue)
-			if rsp.Cause.CauseValue ==
-				pfcpType.CauseNoEstablishedPfcpAssociation {
-				SetUpfInactive(*rsp.NodeID, msg.PfcpMessage.Header.MessageType)
+			smContext.SBIPFCPCommunicationChan <- context.SessionEstablishFailed
+			smContext.SubPfcpLog.Errorf("PFCP Session Establishment rejected with cause [%v]", causeValue)
+			if causeValue == ie.CauseNoEstablishedPFCPAssociation {
+				rspNodeIDStr, err := rsp.NodeID.NodeID()
+				if err != nil {
+					logger.PfcpLog.Errorf("pfcp session establishment response NodeID error: %v", err)
+					return
+				}
+				rspNodeID := context.NewNodeID(rspNodeIDStr)
+				SetUpfInactive(*rspNodeID)
```

Comment:
The msgType argument was not used inside of the `SetUpfInactive` func:

Reference:
- https://github.com/omec-project/smf/blob/master/pfcp/adapter/adapter.go#L323

---
As for having 1 instead of 2 SetUpfInactive funcs, I would argue to tackle this in a separate PR. This one is already huge.

Reply:
Sure, no problem!